docs/courses/robotics/3.5-computer-vision.md

```diff
@@ -0,0 +1,393 @@
+Prints spatial information (position and rotation)
+
+```python
+import cv2
+import numpy as np
+from pupil_apriltags import Detector
+
+def detect_apriltags(image):
+    # Convert to grayscale
+    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+    
+    # Initialize AprilTag detector
+    at_detector = Detector(
+        families="tag36h11",
+        nthreads=1,
+        quad_decimate=1.0,
+        quad_sigma=0.0,
+        refine_edges=1,
+        decode_sharpening=0.25,
+        debug=0
+    )
+    
+    # Detect AprilTags
+    tags = at_detector.detect(gray)
+    
+    return tags
+
+def spatial_referencing(image, tags, tag_size=0.05):
+    # Camera matrix (example values, replace with your camera's parameters)
+    fx, fy, cx, cy = 1000, 1000, image.shape[1]/2, image.shape[0]/2
+    camera_params = [fx, fy, cx, cy]
+    
+    for tag in tags:
+        # Get tag pose
+        pose, e0, e1 = tag.fit_pose(camera_params, tag_size)
+        
+        # Extract rotation and translation
+        rotation_matrix = pose[:3, :3]
+        translation_vector = pose[:3, 3]
+        
+        # Convert rotation matrix to Euler angles
+        euler_angles = cv2.Rodrigues(rotation_matrix)[0].flatten()
+        
+        # Draw tag outline
+        cv2.polylines(image, [np.int32(tag.corners)], True, (0, 255, 0), 2)
+        
+        # Draw tag center and ID
+        center = tuple(map(int, tag.center))
+        cv2.circle(image, center, 5, (0, 0, 255), -1)
+        cv2.putText(image, str(tag.tag_id), (center[0] - 10, center[1] - 10),
+                    cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 255), 2)
+        
+        # Print spatial information
+        print(f"Tag ID: {tag.tag_id}")
+        print(f"Position: X={translation_vector[0]:.2f}, Y={translation_vector[1]:.2f}, Z={translation_vector[2]:.2f}")
+        print(f"Rotation: Roll={np.degrees(euler_angles[0]):.2f}, Pitch={np.degrees(euler_angles[1]):.2f}, Yaw={np.degrees(euler_angles[2]):.2f}")
+        print("---")
+    
+    return image
+
+def main():
+    # Initialize camera
+    cap = cv2.VideoCapture(0)  # Use 0 for default camera, adjust if needed
+    
+    while True:
+        ret, frame = cap.read()
+        if not ret:
+            print("Failed to capture image")
+            break
+        
+        # Detect AprilTags
+        tags = detect_apriltags(frame)
+        
+        # Perform spatial referencing
+        frame_with_tags = spatial_referencing(frame, tags)
+        
+        # Display the result
+        cv2.imshow('AprilTag Spatial Referencing', frame_with_tags)
+        
+        if cv2.waitKey(1) & 0xFF == ord('q'):
+            break
+    
+    cap.release()
+    cv2.destroyAllWindows()
+
+if __name__ == "__main__":
+    main()
+```
+
+This example assumes you have a camera connected and AprilTags in view. You'll need to adjust the camera matrix parameters (fx, fy, cx, cy) to match your specific camera's calibration. Also, make sure to set the correct tag_size in meters.
```

Comment:
Does the video go over how to set calibration parameters? If not, no worries. Could you include a link or provide some instructions about going through the calibration process?

Reply:
Maybe also mention as optional a ruler